### PR TITLE
Improve number formatting in tensor print

### LINKED ISSUE #6768: Tensor printing should display requires_grad
I think one or more of the following would be nice:
- tensor printing displays requires_grad
- tensor printing displays "if this tensor is a leaf node" (but find a better name for this). Most of the times when I want to know if my tensor requires_grad I want to know if it gradients will accumulate in this.

cc @li-roy @gchanan @SsnL

### LINKED ISSUE #6811: repr of integer tensors with large values doesn't match construction
At least in python 3 (I didn't check python 2), literals in scientific notation are always floats.

So:
```
>>> a=torch.tensor(425245252)
>>> a.dtype
torch.int64
>>> a
tensor(4.2525e+08)
>>> torch.tensor(4.2525e+08).dtype
torch.float32
```

CC @li-roy

### LINKED ISSUE #6865: [PyTorch] Don't use scientific notation for printing integer tensors
We currently use scientific notation (E-notation) to print large integer tensors. We shouldn't do this because it hides meaningful digits and because E-notation corresponds to floating point data types in Python.

```python
>>> x = torch.ones(525076, dtype=torch.int)
>>> x.sum()
tensor(5.2508e+05)
>>> x.sum().item()
525076
```

### LINKED ISSUE #6918: repr of tensors seems to have extra space
- PyTorch or Caffe2: Pytorch
- How you installed PyTorch (conda, pip, source): pip
- OS: macos 10.13
- PyTorch version: 0.4.0
- Python version: 3.6.5

If I run this code in the interpreter,

```python
import torch
t = torch.tensor([1, 2])
repr(t)
```

I get this output:

```python
tensor([~1,~~2])
```

I put a `~` instead of a space for clarity. As you can see, extra spaces are printed. This happens even if I create a singleton 1D tensor.

### LINKED ISSUE #7213: Get rid of scaling factor in print
Suggested in #7175. Creating new issue to track.

### PR DESCRIPTION
**Pull request by li-roy**
- Integer tensors won't be printed in scientific notation
- Removed scaling factor for all tensors
- Only leave spaces for negative signs if one of the printed elements is going to have an negative sign

Fixes #7213, fixes #6918, fixes #6865, fixes #6811, fixes #6768.

Examples
```
>>> torch.tensor(2341234123412341)
tensor(2341234123412341)

>>> torch.tensor([1e28, 1e-28])
tensor([1.0000e+28, 1.0000e-28])

>>> torch.tensor([1e28, -1e-28])
tensor([ 1.0000e+28, -1.0000e-28])

>>> torch.tensor([1, 2])
tensor([1, 2])

>>> torch.tensor([1, -2])
tensor([ 1, -2])

>>> torch.randn(1001)
tensor([-0.5033, -0.4142,  0.7961,  ..., -0.9400, -0.4431, -0.6192])

>>> torch.randn(20)
tensor([ 0.8574, -1.1225,  0.3135,  1.7848, -0.8342,  0.1214, -0.6393, -0.3480,
         1.4114,  0.5102, -0.8206,  1.6022,  0.6310,  1.9507, -1.5694,  1.3168,
         0.1103,  0.5242,  0.3141, -0.0692])

>>> torch.randn(5, dtype=torch.double)
tensor([-1.2961,  0.6950,  0.1357, -0.6473,  0.2605], dtype=torch.float64)
```

### CODE REVIEW

**Comment by SsnL** [comment]
Cool! Can you post some example outputs?

File: torch/_tensor_str.py [hunk 1]
```diff
@@ -65,132 +65,128 @@ def set_printoptions(
...
 
 
-def _get_min_log_scale():
-    min_positive = float_info.min * float_info.epsilon  # get smallest denormal
-    if min_positive == 0:  # use smallest normal if DAZ/FTZ is set
-        min_positive = float_info.min
-    return math.ceil(math.log(min_positive, 10))
+class _Formatter(object):
+    def __init__(self, tensor):
+        tensor = tensor.view(tensor.nelement())
```
Comment (gchanan) [inline torch/_tensor_str.py]: this isn't right, see the failing tests (you have to make the tensor contiguous).  What's the advantage of this over the old code given you may have to make the entire tensor contiguous?

File: torch/_tensor_str.py [hunk 2]
```diff
@@ -65,132 +65,128 @@ def set_printoptions(
...
-    # The leading space for positives is ugly on scalars, so we strip it
-    return scalar_str.lstrip()
+def _scalar_str(self, formatter):
+    return formatter.format(self.item())
 
 
-def _vector_str(self, indent, fmt, scale, sz, summarize):
-    element_length = sz + 3
+def _vector_str(self, indent, formatter, summarize):
+    element_length = formatter.width() + 2
```
Comment (gchanan) [inline torch/_tensor_str.py]: can you comment what the magic number here is supposed to be for (I realize it was there before, but I don't remember what is was for before and I don't know what it's for now).

File: torch/_tensor_str.py [hunk 3]
```diff
@@ -65,132 +65,128 @@ def set_printoptions(
...
+        self.max_width = 1
+
+        if self.floating_dtype:
+            for value in tensor.tolist():
+                value_str = '{}'.format(value)
+                self.max_width = max(self.max_width, len(value_str))
 
+        else:
+            try:
+                for value in tensor.tolist():
```
Comment (gchanan) [inline torch/_tensor_str.py]: this looks strange -- you are calling ceil when floating_dtype is false but not when it's true?  Am I missing something here?

File: torch/_tensor_str.py (hunk 1 above)
Comment (li-roy) [reply torch/_tensor_str.py]: The old code makes a contiguous copy anyways. (And so will the new code) The main thing the new method is doing is making sure we're setting the item width to be exactly as large as we need. For example, we want `[1, 2]` and not `[ 1,  2]` if no elements are negative. Actually attempting to format all the numbers was the best way I could think of dealing with cases like negative 0.

**Comment by fmassa** [comment]
I think the prints now are a bit verbose.
Can't we print the dtype/device/requires_grad if they are not what we would expect to be default?

Also, quite a few people wanted to have a shape information in the print. While this would break the constructor-valid contract, it could be worth considering.

**Comment by li-roy** [comment]
That's what it was like before, but there was some discussion on #6912 on whether it makes sense to do that, seeing as those defaults can be changed.

**Comment by fmassa** [comment]
I personally would only print requires_grad if it's True. Same thing for device, only print it if it's not on the CPU.

Also, I think that at some point this becomes a matter of taste, and maybe adding options to the printoptions to always enable / disable certain prints might be worth considering.

**Comment by alok** [comment]
I don't think there should be a space before the first element of a tensor, since numpy and python lists don't do that

**Comment by li-roy** [comment]
@alok Numpy does when it's leaving space for a negative.

```
>>> np.array([1,-2])
array([ 1, -2])
```

**Comment by ezyang** [comment]
So, what does the new formatting look like? An expect test for a simple tensor would go a long way towards making it more obvious to reviewers.

File: torch/_tensor_str.py [hunk 4]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
 
+        else:
+            copy = torch.DoubleTensor(tensor.size()).copy_(tensor).view(tensor.nelement())
+            try:
+                for value in copy.tolist():
+                    if value != math.ceil(value):
+                        self.int_mode = False
+                        break
+            # nonfinites will throw errors
+            except (ValueError, OverflowError):
```
Comment (gchanan) [inline torch/_tensor_str.py]: just curious: where does the error come from?

File: torch/_tensor_str.py [hunk 5]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
+                self.int_mode = False
+
+            if self.int_mode:
+                for value in copy.tolist():
```
Comment (gchanan) [inline torch/_tensor_str.py]: can we avoid converting to lists multiple times?

File: torch/_tensor_str.py [hunk 6]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
+                    value_str = '{:.0f}'.format(value)
+                    if math.isnan(value) or math.isinf(value):
+                        self.max_width = max(self.max_width, len(value_str))
+                    else:
+                        self.max_width = max(self.max_width, len(value_str) + 1)
```
Comment (gchanan) [inline torch/_tensor_str.py]: what's the +1 for?  Can you comment?

File: torch/_tensor_str.py [hunk 7]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
+    return formatter.format(self.item())
 
 
-def _vector_str(self, indent, fmt, scale, sz, summarize):
-    element_length = sz + 3
+def _vector_str(self, indent, formatter, summarize):
+    # length includes spaces and comma between elements
+    element_length = formatter.width() + 2
     elements_per_line = int(math.floor((PRINT_OPTS.linewidth - indent) / (element_length)))
+    if elements_per_line < 1:
```
Comment (gchanan) [inline torch/_tensor_str.py]: nit: nicer to just put this in a max above.

File: torch/_tensor_str.py (hunk 4 above)
Comment (li-roy) [reply torch/_tensor_str.py]: It happens when I do `math.ceil` on `nan` and `inf`.

**Comment by ezyang** [comment]
FYI, you can use `assertExpected` instead of hard coding the strings, and then pass `--accept` to automatically accept the actual string as the expected result. Saves a bunch of copy pasting.

**Comment by gchanan** [comment]
FWIW I'm not a fan of `assertExpected` here -- these are really simple tests and it's nicer to just have them visible inline.  But it's fine now that it's done.

File: torch/_tensor_str.py [hunk 8]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
+        self.max_width = 1
+
+        if not self.floating_dtype:
+            copy = torch.LongTensor(tensor.size()).copy_(tensor).view(tensor.nelement())
+            for value in copy.tolist():
+                value_str = '{}'.format(value)
+                self.max_width = max(self.max_width, len(value_str))
 
+        else:
+            copy = torch.DoubleTensor(tensor.size()).copy_(tensor).view(tensor.nelement())
```
Comment (gchanan) [inline torch/_tensor_str.py]: nit: you could move these two lines out of the if/else.

File: torch/_tensor_str.py [hunk 9]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
+            except (ValueError, OverflowError):
+                self.int_mode = False
+
+            if self.int_mode:
+                for value in copy_list:
+                    value_str = '{:.0f}'.format(value)
+                    if math.isnan(value) or math.isinf(value):
+                        self.max_width = max(self.max_width, len(value_str))
+                    else:
+                        # for finites, appending a decimal will increase the len by 1
```
Comment (gchanan) [inline torch/_tensor_str.py]: I'm a bit confused about the comment -- we are in int_mode, so how do we know we are going to print a decimal?

File: torch/_tensor_str.py [hunk 10]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
-            format = '{{:{}.{}e}}'.format(sz, prec)
-            fmt_fn = format.format
-        else:
-            if exp_max > prec + 1 or exp_max < 0:
-                sz = max(min_sz, 7)
-                scale = math.pow(10, max(exp_max - 1, _min_log_scale))
-            else:
-                if exp_max == 0:
-                    sz = 7
+                if exp_max - exp_min > PRINT_OPTS.precision or exp_max > 8 or exp_min < -4:
```
Comment (gchanan) [inline torch/_tensor_str.py]: any specific rationale for these numbers?

File: torch/_tensor_str.py [hunk 11]
```diff
@@ -65,132 +65,130 @@ def set_printoptions(
...
+        if self.floating_dtype:
+            if self.int_mode:
+                ret = '{:.0f}'.format(value)
+                if not (math.isinf(value) or math.isnan(value)):
+                    ret += '.'
+            elif self.sci_mode:
+                ret = ('{:' + str(self.max_width) +
+                       '.' + str(PRINT_OPTS.precision) + 'e}').format(value)
+            else:
+                ret = ('{:.' + str(PRINT_OPTS.precision) + 'f}').format(value)
```
Comment (gchanan) [inline torch/_tensor_str.py]: nit: you have the actual formats multiple times -- can you have them call into a common function?

**Review by gchanan — APPROVED** [review]
I have some small comments / suggestions, but I think this is good to go.

File: torch/_tensor_str.py (hunk 11 above)
Comment (li-roy) [reply torch/_tensor_str.py]: Actually only 2 of the 4 are the same, but I'll change the concat to a format to make it clearer.

File: torch/_tensor_str.py (hunk 8 above)
Comment (li-roy) [reply torch/_tensor_str.py]: In the if statement I'm actually making a LongTensor instead of a Double tensor, so the two statements are different.

File: torch/_tensor_str.py (hunk 10 above)
Comment (gchanan) [reply torch/_tensor_str.py]: that's what I thought -- a comment to that effect would be helpful in case we forget.

File: torch/_tensor_str.py (hunk 8 above)
Comment (gchanan) [reply torch/_tensor_str.py]: ah, right, my mistake.  It still might be clearer if you just passed a dtype (based on int_mode) to torch.empty, but this is nit picky.
